SignalServiceKit/src/Messages/OWSMessageReceiver.m

```diff
@@ -0,0 +1,338 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSMessageReceiver.h"
+#import "OWSSignalServiceProtos.pb.h"
+#import "TSDatabaseView.h"
+#import "TSMessagesManager.h"
+#import "TSStorageManager.h"
+#import "TSYapDatabaseObject.h"
+#import <YapDatabase/YapDatabaseConnection.h>
+#import <YapDatabase/YapDatabaseTransaction.h>
+#import <YapDatabase/YapDatabaseView.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+#pragma mark - Persisted data model
+@class OWSSignalServiceProtosEnvelope;
+
+@interface OWSMessageProcessingJob : TSYapDatabaseObject
+
+@property (nonatomic, readonly) OWSSignalServiceProtosEnvelope *envelopeProto;
+@property (nonatomic, readonly) NSDate *createdAt;
```

Comment:
Should we use this value to initialize `TSMessage.receivedAtTimestamp`?  

* That'd make the "received" time a bit more accurate.
* It'd preserve ordering for messages we decrypt "later", e.g. after receiving a response to a group info update.  That might be desirable or not.
* It might have some surprisingly side effects in other cases where we process the queue out of received order.